src/libAtomVM/nifs.c

```diff
@@ -3505,6 +3557,374 @@ static term nif_atomvm_read_priv(Context *ctx, int argc, term argv[])
+        fd = open(path, posix_flags, mode);
+    } else {
+        fd = open(path, posix_flags);
+    }
+    free((void *) path);
+    if (fd < 0) {
+        // Return an error.
+        if (UNLIKELY(memory_ensure_free_opt(ctx, TUPLE_SIZE(2), MEMORY_CAN_SHRINK) != MEMORY_GC_OK)) {
+            RAISE_ERROR(OUT_OF_MEMORY_ATOM);
+        }
+        result = term_alloc_tuple(2, &ctx->heap);
+        term_put_tuple_element(result, 0, ERROR_ATOM);
+        term_put_tuple_element(result, 1, posix_errno_to_atom(errno, glb));
+    } else {
+        // Return a resource object
+        struct PosixFd *fd_obj = enif_alloc_resource(glb->posix_fd_resource_type, sizeof(struct PosixFd));
+        fd_obj->fd = fd;
+        if (UNLIKELY(memory_ensure_free_opt(ctx, TUPLE_SIZE(2) + TERM_BOXED_RESOURCE_SIZE, MEMORY_CAN_SHRINK) != MEMORY_GC_OK)) {
+            RAISE_ERROR(OUT_OF_MEMORY_ATOM);
+        }
+        term obj = term_from_resource(fd_obj, &ctx->heap);
+        enif_release_resource(fd_obj);
+        result = term_alloc_tuple(2, &ctx->heap);
+        term_put_tuple_element(result, 0, OK_ATOM);
+        term_put_tuple_element(result, 1, obj);
+    }
+
+    return result;
+}
+
+static term nif_atomvm_posix_close(Context *ctx, int argc, term argv[])
+{
+    UNUSED(argc);
+
+    void *fd_obj_ptr;
+    if (UNLIKELY(!enif_get_resource(erl_nif_env_from_context(ctx), argv[0], ctx->global->posix_fd_resource_type, &fd_obj_ptr))) {
+        RAISE_ERROR(BADARG_ATOM);
+    }
+    struct PosixFd *fd_obj = (struct PosixFd *) fd_obj_ptr;
+    if (fd_obj->fd) {
+        close(fd_obj->fd);
+        fd_obj->fd = 0;
+    }
+
+    return OK_ATOM;
+}
+
+static term nif_atomvm_posix_read(Context *ctx, int argc, term argv[])
+{
+    GlobalContext *glb = ctx->global;
+    UNUSED(argc);
+    term count_term = argv[1];
+    VALIDATE_VALUE(count_term, term_is_integer);
+    int count = term_to_int(count_term);
+
+    void *fd_obj_ptr;
+    if (UNLIKELY(!enif_get_resource(erl_nif_env_from_context(ctx), argv[0], glb->posix_fd_resource_type, &fd_obj_ptr))) {
+        RAISE_ERROR(BADARG_ATOM);
+    }
+    struct PosixFd *fd_obj = (struct PosixFd *) fd_obj_ptr;
+    if (UNLIKELY(memory_ensure_free_opt(ctx, term_binary_data_size_in_terms(count) + BINARY_HEADER_SIZE + TERM_BOXED_SUB_BINARY_SIZE + TUPLE_SIZE(2), MEMORY_CAN_SHRINK) != MEMORY_GC_OK)) {
+        RAISE_ERROR(OUT_OF_MEMORY_ATOM);
+    }
+    term bin_term = term_create_uninitialized_binary(count, &ctx->heap, glb);
+    int res = read(fd_obj->fd, (void *) term_binary_data(bin_term), count);
+    if (res < 0) {
+        // Return an error.
+        if (UNLIKELY(memory_ensure_free_opt(ctx, TUPLE_SIZE(2), MEMORY_CAN_SHRINK) != MEMORY_GC_OK)) {
+            RAISE_ERROR(OUT_OF_MEMORY_ATOM);
+        }
+        term ret = term_alloc_tuple(2, &ctx->heap);
+        term_put_tuple_element(ret, 0, ERROR_ATOM);
+        term_put_tuple_element(ret, 1, posix_errno_to_atom(errno, glb));
+        return ret;
+    }
+    if (res == 0) {
```

Comment:
Opinion: let's group them using `else`, } else if (res == 0) {`

Reply:
I beg to differ as we have return statements.